### PR TITLE
Handle pagination of http resources

### PR DESCRIPTION
**Pull request by samsteyer**
Let's try this again!

This pull request adds a boolean `paginate` option to the http source. If `True`, this option tells `landgrab` to get data from multiple pages of paginated http source and combine them into one .jsonlines output.

### CODE REVIEW

File: landgrab/source/http.py [hunk 1]
```diff
@@ -8,20 +11,69 @@ class HTTPSource(BaseSource):
...
+                        r = requests.get(
+                            self.uri,
+                            params=query_params
+                        )
+                    if r.status_code < 400:
+                        results = r.json()
+                        for tariff in results[self.results_key]:
+                            writer.write(tariff)
+        else:
+            urllib.urlretrieve(self.uri, self.f.name)
```
Comment (jerluc) [inline landgrab/source/http.py]: If we're importing requests, we might as well use it here as well for getting the data, that way we get support for multiple methods, auth headers, etc.

File: landgrab/source/http.py [hunk 2]
```diff
@@ -8,20 +11,69 @@ class HTTPSource(BaseSource):
...
         return self
 
+    def _query_param_generator(self):
```
Comment (jerluc) [inline landgrab/source/http.py]: Might want to rename to something like `_pagination_query_param_generator`

File: landgrab/source/http.py [hunk 3]
```diff
@@ -8,20 +11,69 @@ class HTTPSource(BaseSource):
...
+        page_index = self.current_page
+        for i in range(self.max_pages):
+            query_params = {
+                self.results_per_page_param: self.results_per_page,
+                self.page_index_param: page_index
+            }
+            query_params.update(self.query_params)
+            yield query_params
```
Comment (jerluc) [inline landgrab/source/http.py]: Neat

File: landgrab/source/http.py [hunk 4]
```diff
@@ -8,20 +11,69 @@ class HTTPSource(BaseSource):
...
+            page_index += self.results_per_page
```
Comment (jerluc) [inline landgrab/source/http.py]: This is the page offset, not the index, right?

File: requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 click
+jsonlines
```
Comment (jerluc) [inline requirements.txt]: What additional does this package provide that `my_file.write(json.dumps(obj))` doesn't?

File: landgrab/source/http.py [hunk 6]
```diff
@@ -8,20 +11,69 @@ class HTTPSource(BaseSource):
...
     """
-    def __init__(self, uri, method='GET'):
+    def __init__(self, uri, method='GET', paginate=False, results_per_page_param=None,
+                 results_per_page=None, page_index_param=None, max_pages=None, current_page=0,
+                 query_params=None, results_key='results', username=None, password=None):
         self.uri = uri
         self.method = method
+        self.paginate = paginate
+
+        # Vars only necessary if paginate is `True`
```
Comment (jerluc) [inline landgrab/source/http.py]: You may want to consider writing it such that the `paginate` option is not a `bool`, but rather a `dict` containing these options. Then there is no conditional here, and it makes it more obvious that these properties are only required if one is trying to paginate results.

Comment (samsteyer) [reply landgrab/source/http.py]: Cool, change made

File: landgrab/source/http.py (hunk 1 above)
Comment (samsteyer) [reply landgrab/source/http.py]: ✔️

File: landgrab/source/http.py (hunk 2 above)
Comment (samsteyer) [reply landgrab/source/http.py]: ✔️

File: landgrab/source/http.py (hunk 3 above)
Comment (samsteyer) [reply landgrab/source/http.py]: 😺

File: landgrab/source/http.py (hunk 4 above)
Comment (samsteyer) [reply landgrab/source/http.py]: Yes, good point, changed.

File: requirements.txt (hunk 5 above)
Comment (samsteyer) [reply requirements.txt]: None - removing it

Comment (samsteyer) [reply requirements.txt]: Is `'\n'` the right character for returns in .jsonlines files?

**Pull request merged**